### PR TITLE
support repl

### PR DESCRIPTION
**Pull request by ahaoboy**
### Description of changes

Adding a simple repl mode

https://nodejs.org/en/learn/command-line/how-to-use-the-nodejs-repl

https://github.com/DelSkayn/rquickjs/blob/master/examples/rquickjs-cli/src/main.rs

### Checklist

- [ ] Created unit tests in `tests/unit` and/or in Rust for my feature if needed
- [ ] Ran `make fix` to format JS and apply Clippy auto fixes
- [ ] Made sure my code didn't add any additional warnings: `make check`
- [ ] Added relevant type info in `types/` directory
- [ ] Updated documentation if needed ([API.md](API.md)/[README.md](README.md)/Other)

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: llrt_core/src/vm.rs [hunk 1]
```diff
@@ -184,6 +184,28 @@ impl Vm {
 
         self.run(source, strict, global).await;
     }
+
+    pub async fn run_repl(&self) {
+        self.ctx
+            .with(|ctx| -> Result<()> {
+                let global = ctx.globals();
+                let console: Object = global.get("console")?;
+                let js_log: Function = console.get("log")?;
```
Comment (richarddavison) [inline llrt_core/src/vm.rs:193]: We can call these methods directly from rust and print. Check the console module.

File: llrt_core/src/vm.rs [hunk 2]
```diff
@@ -184,6 +184,28 @@ impl Vm {
...
+                loop {
+                    let mut input = String::new();
```
Comment (richarddavison) [inline llrt_core/src/vm.rs:195]: We don't need to create a new string very time. We can create one outside the loop and clear it.

File: llrt_core/src/vm.rs [hunk 3]
```diff
@@ -184,6 +184,28 @@ impl Vm {
...
+                    print!("> ");
```
Comment (richarddavison) [inline llrt_core/src/vm.rs:196]: Maybe we also should print the version header? (see version method).

File: llrt_core/src/vm.rs [hunk 4]
```diff
@@ -184,6 +184,28 @@ impl Vm {
...
+                    std::io::Write::flush(&mut std::io::stdout())?;
+                    std::io::stdin().read_line(&mut input)?;
+                    ctx.eval::<Value, _>(input.as_bytes())
+                        .and_then(|ret| js_log.call::<(Value<'_>,), ()>((ret,)))
+                        .catch(&ctx)
+                        .unwrap_or_else(|err| println!("{err}"));
```
Comment (richarddavison) [inline llrt_core/src/vm.rs:202]: I think the right approach here would be to evaluate the code and return value and pass that value to format method and write to std out

**Review by richarddavison — CHANGES_REQUESTED** [review]
Cool idea, I like it! Some minor comment comments!

File: llrt_core/src/vm.rs [hunk 5]
```diff
@@ -184,6 +185,44 @@ impl Vm {
...
+        self.ctx
+            .with(|ctx| -> Result<()> {
+                let mut input = String::new();
+                println!("Welcome to llrt v{}", VERSION);
+                loop {
+                    print!("> ");
+                    std::io::Write::flush(&mut std::io::stdout())?;
+                    std::io::stdin().read_line(&mut input)?;
+                    match ctx.eval::<Value, _>(input.as_bytes()) {
+                        Ok(v) => console::log_std_out(
```
Comment (richarddavison) [inline llrt_core/src/vm.rs:199]: Call `pub fn format` in console instead to get the formated string back. Then do `println` or `eprintln` (for exception) with that string.

**Review by richarddavison — CHANGES_REQUESTED** [review]
Almost there! Just a minor change

File: llrt_core/src/vm.rs [hunk 6]
```diff
@@ -184,6 +185,48 @@ impl Vm {
...
+    pub async fn run_repl(&self) {
+        self.ctx
+            .with(|ctx| -> Result<()> {
+                let mut input = String::new();
+                println!("Welcome to llrt v{}", VERSION);
+                loop {
+                    print!("> ");
+                    std::io::Write::flush(&mut std::io::stdout())?;
+                    std::io::stdin().read_line(&mut input)?;
+                    match ctx.eval::<Value, _>(input.as_bytes()) {
```
Comment (richarddavison) [inline llrt_core/src/vm.rs:198]: This still needs a bit of work. Something like this (pseudo code):
```rust
match ctx.eval::<Value, _>(input.as_bytes()).map(|v|console::format(&ctx, Rest(vec![v])).catch(&ctx){
  Ok(s) => println!("{s}"),
  Err(caught_err) => {
        match caught_err.into_value(ctx).map(|v|console.format(&ctx, Rest(vec![v]))){
           Ok(s) => eprintln!("Error: {s}"),
           Err(_) => eprintln!("Error: {:?}", caught_err)
        }
  }
}
```

**Review by richarddavison — CHANGES_REQUESTED** [review]
Some suggestions to simplify. Would also be great if we can extract the repl inner logic to an own file and pass input and output to it. That way we can make it testable from rust.

**Comment by ahaoboy** [comment]
> That way we can make it testable from rust.

I honestly have no idea how to test the repl, simulate keyboard input?

**Comment by richarddavison** [comment]
> > That way we can make it testable from rust.
> 
> I honestly have no idea how to test the repl, simulate keyboard input?

No this can be simpler. Extract the processing of the line and just test that. We don't need to test capturing of std input.
In order to this we should probably move the entire repl capability from vm.rs and into a separate file (repl.rs). In there we can have like:
```rust
fn process_line(&line) -> std::result::Result<String,String>{
 //existing line processing logic but return strings for error and console rather than printing them stderr or stdout
}
```

And then we can test `process_line` with different input valid and invalid :) Hope it makes sense.

**Review by richarddavison — APPROVED** [review]
Nice thank you!

**Pull request merged**